examples/pytorch/ogb/directional_GSN/preprocessing.py

```diff
@@ -0,0 +1,285 @@
+from ogb.graphproppred import DglGraphPropPredDataset, Evaluator
+import torch
+import numpy as np
+import networkx as nx
+import graph_tool as gt
+import graph_tool.topology as gt_topology
+from torch.utils.data import Dataset
+import types
+import time
+from tqdm import tqdm
+import dgl
+import os
+
+def to_undirected(edge_index):
+    row, col = edge_index
+    row, col = torch.cat([row, col], dim=0), torch.cat([col, row], dim=0)
+    edge_index = torch.stack([row, col], dim=0)
+
+    return edge_index
+
+def automorphism_orbits(edge_list):
+    
+    ##### vertex automorphism orbits ##### 
+    graph = gt.Graph(directed=False)
+    graph.add_edge_list(edge_list)
+    gt.stats.remove_self_loops(graph)
+    gt.stats.remove_parallel_edges(graph)
+
+    # compute the vertex automorphism group
+    aut_group = gt_topology.subgraph_isomorphism(graph, graph, induced=False, subgraph=True, generator=False)
+
+    orbit_membership = {}
+    for v in graph.get_vertices():
+        orbit_membership[v] = v
+    
+    # whenever two nodes can be mapped via some automorphism, they are assigned the same orbit
+    for aut in aut_group:
+        for original, vertex in enumerate(aut):
+            role = min(original, orbit_membership[vertex])
+            orbit_membership[vertex] = role
+        
+    orbit_membership_list = [[],[]]
+    for vertex, om_curr in orbit_membership.items():
+        orbit_membership_list[0].append(vertex)
+        orbit_membership_list[1].append(om_curr)
+
+    # make orbit list contiguous (i.e. 0,1,2,...O)
+    _, contiguous_orbit_membership = np.unique(orbit_membership_list[1], return_inverse = True)
+
+    orbit_membership = {vertex: contiguous_orbit_membership[i] for i,vertex in enumerate(orbit_membership_list[0])}
+
+    orbit_partition = {}
+    for vertex, orbit in orbit_membership.items():
+        orbit_partition[orbit] = [vertex] if orbit not in orbit_partition else orbit_partition[orbit]+[vertex]
+    
+    aut_count = len(aut_group)
+
+    return graph, orbit_partition, orbit_membership, aut_count
+
+def induced_edge_automorphism_orbits(edge_list):
+    
+    ##### induced edge automorphism orbits (according to the vertex automorphism group) #####
+    graph, orbit_partition, orbit_membership, aut_count = automorphism_orbits(edge_list=edge_list)
+    edge_orbit_partition = dict()
+    edge_orbit_membership = dict()
+    edge_orbits2inds = dict()
+    ind = 0
+    
+    edge_list = to_undirected(torch.tensor(graph.get_edges()).transpose(1,0)).transpose(1,0).tolist()
```

Comment:
To perform `.transpose(1,0)` twice is quite ugly. Perhaps we should directly modify `to_undirected` instead.